### ISSUE TITLE
Drop, clone unsafe and theading races

### ISSUE THREAD

**Issue by karlri** [stevedonovan/mosquitto-client#5]
Nice initiative to provide a Rust interface to libmosquitto, I like it, however, the way the code is structured does not quite fit Rust design goals, it is unsafe without showing it in the interface. May I point out 2 issues and suggest possible fixes?

1. drop and clone are unsafe, because the clone will not free the underlying memory, and the original will. If the clone outlives the original, then the memory is freed.
Suggested fix: remove the clone implementation and owned attribute, let people use Rc if they need multiple references.

2. the documentation for the c code states that it the c code is only thread safe if a threading enable function is called before connect, however the wrapping rust type implements Send and Sync. This can create various race conditions and crashes if used in several threads.
Suggested fix: Break out functions to a trait Mosquitto, make 2 different structs with the same implementation, one for single thread, which is Clone but neither Send nor Sync, and one for multi thread which is both Send and Sync and Clone, and that one could call set_threaded in constructor.

It would be great if these issues could be adressed. The above changes are just a suggestion :) Unfortunately, I don't have time to submit a PR.

**Comment by stevedonovan** [comment]
Yeah, I did it in my early Rust days - first unsafe code I did, and it shows!

TIme to revisit...
